### PR TITLE
Add test coverage for integration with Solidus Starter Frontend

### LINKED ISSUE #166: Add test coverage for the integration with `solidus_starter_frontend`
## What 

QA the frontend integration, make sure all the necessary partials and behaviors are there and work as expected.
Add missing specs to cover any regular or edge case.

```[tasklist]
### Test cases
- [ ] checking out as a guest
- [ ] checking out as a registered user
- [ ] checking with every test card provided by stripe (cover most edge cases)
- [ ] checking out with a saved payment source
- [ ] checking out with/without autocapture
```

## Why

It's likely that once finished we'll still have some edge cases to cover, this is meant as a final QA that should be done before releasing.

### LINKED ISSUE #187: Not possible to complete checkout after failure
**Steps to reproduce**

- Proceed with a checkout until Stripe's payment form is rendered.
- Use a [card that will be declined](https://stripe.com/docs/testing#declined-payments) (e.g., `4000000000000002`).
- If #184 has been done, do nothing. Otherwise, fail the payment from the backend from a private browser window.
- Go back to the form and now try a [valid card](https://stripe.com/docs/testing#cards) (e.g., `4242424242424242`).

**Result**
Solidus fails with `Cannot transition state via :pend from :failed (Reason(s): State cannot transition via "pend")`, but the payment is authorized on Stripe.

**Expected result**
Solidus checkout process is able to go forward.

**Considerations**
Maybe we could handle that with the `payment_intent.amount_capturable_updated` webhook event that we receive, and "unfail" a payment in that case.

### PR DESCRIPTION
**Pull request by rainerdema**
## Summary

Closes #166

This pull request adds improved test coverage for using the gem with [solidus_starter_frontend](https://github.com/solidusio/solidus_starter_frontend).
The `checkout_spec.rb` has been refactored to include specific use cases with the addition of new scenarios and reorganized test blocks.

The tests cover the following cases:

| Intents flow| Description| Setup Feature Usage | Skip Confirm Step | User | Guest | Payment Result|
|-:|:-|:-:|:-:|:-:|:-:|:-|
| Setup Intent | Credit card with payment capture and wallet reuse | all | no| yes| yes | Successfully created/processed/reused |
| Setup Intent | 3D secure card without capture | all | no| yes | | Required action present |
| Payment Intent | Credit card with payment capture and wallet reuse | all | yes/no | yes| yes | Successfully created/processed/reused |
| Payment Intent | 3D secure card with payment capture | all| yes | yes | | Successfully created/processed |
| Setup Intent | 3D secure card authentication failure | - | -| yes | | Failed authentication |
| Setup Intent | Transaction failure for incorrect data | - | -| yes | | Failed transaction |
| Setup Intent| Incomplete data form submission failure | - | -| yes | | Failed submission |

These tests help ensure the reliability and stability of the gem when used with [solidus_starter_frontend](https://github.com/solidusio/solidus_starter_frontend).

Closes also #187

This PR includes tests that cover all the cards mentioned in the issue ([card that will be declined](https://stripe.com/docs/testing#declined-payments)). 
The issues mentioned in the ticket may be related to previous situations addressed in other PRs:
<details>
    <summary>Refer to the video below to see the current situation in a test environment</summary>
   
  <video src="https://user-images.githubusercontent.com/19948291/224904672-36324941-0a5f-4b2d-9a5a-3049b376cbb7.mov"/>
</details>

## Checklist

Check out our [PR guidelines](https://github.com/solidusio/.github/blob/master/CONTRIBUTING.md#pull-request-guidelines) for more details.

The following are mandatory for all PRs:

- [x] I have written a thorough PR description.
- [x] I have kept my commits small and atomic.
- [x] [I have used clear, explanatory commit messages](https://github.com/solidusio/.github/blob/main/CONTRIBUTING.md#writing-good-commit-messages).

The following are not always needed:

- 📖 I have updated the README to account for my changes.
- 📑 I have documented new code [with YARD](https://www.rubydoc.info/gems/yard/file/docs/Tags.md).
- 🛣️ I have opened a PR to update the [guides](https://github.com/solidusio/edgeguides).
- ✅ I have added automated tests to cover my changes.
- 📸 I have attached screenshots to demo visual changes.

## Reviewer Guidelines
To better review the changes made to `checkout_spec.rb`, I suggest reviewing them on the branch rather than the diff.

Here is a link to the relevant file:
[solidus_stripe/spec/system/frontend/solidus_stripe/checkout_spec.rb](https://github.com/solidusio/solidus_stripe/blob/nebulab/rainerd/add-test-coverag-for-the-integration-with-solidus-starter-frontend/spec/system/frontend/solidus_stripe/checkout_spec.rb)
(I can split it into multiple commits if needed)

### CODE REVIEW

**Comment by rainerdema** [comment]
Please let me know if it would be preferred to split the tests into multiple commits 🙏

File: spec/system/frontend/solidus_stripe/checkout_spec.rb [hunk 1]
```diff
@@ -2,218 +2,302 @@
 
 require 'solidus_stripe_spec_helper'
 
-RSpec.describe "Checkout with Stripe", :js do
+RSpec.describe 'Checkout', :js do
```
Comment (elia) [inline spec/system/frontend/solidus_stripe/checkout_spec.rb]: If we're matching the file name maybe we can go one step further and use "SolidusStripe Checkout"?
No big deal, but it's easier to understand where is the problem in case of failures.

File: spec/system/frontend/solidus_stripe/checkout_spec.rb [hunk 2]
```diff
@@ -2,218 +2,302 @@
...
-    it "completes as a registered user using payment intents #{skip_confirm || ' not'} skipping confirm" do
-      create(
-        :stripe_payment_method,
-        preferred_stripe_intents_flow: 'payment',
-        preferred_skip_confirmation_for_payment_intent: skip_confirm,
-      )
-      visit_payment_step(user: create(:user))
-      choose_new_stripe_payment
-      fill_stripe_form
+        it 'successfully reuses a previously saved card from the user\'s wallet' do
```
Comment (elia) [inline spec/system/frontend/solidus_stripe/checkout_spec.rb]: nit: if rubocop is not allowing to use double quotes when we have a single quote inside the string we should change the rule

File: spec/system/frontend/solidus_stripe/checkout_spec.rb [hunk 3]
```diff
@@ -2,218 +2,302 @@
...
+          # Expect the setup intent to be created successfully and complete the order
+          expect_setup_intent_to_be_created_successfully
+          complete_order
+          expect(page).to have_content('Your order has been processed successfully')
+
+          # Note that, in case of a Setup Intent, even if it is authorized, the payment intent
+          # cannot be captured because a required action is needed most of the times.
+          expect_payment_intent_to_be_created_with_required_action
+
+          # TODO: These cards cannot be reused as the required action is not handled.
```
Comment (elia) [inline spec/system/frontend/solidus_stripe/checkout_spec.rb]: nit: a failing test preceded by a `pending` would be ok here too, not blocking for the approval but if it's super easy to add would be of great help for the next one needing to implement the solution.

**Review by elia — APPROVED** [review]
🙌 awesome!

I left a couple of nits for your consideration, they're not blocking tho.

**Review by kennyadsl — APPROVED** [review]
Lovely, thanks @rainerdema!

File: spec/support/solidus_stripe/checkout_test_helper.rb [hunk 4]
```diff
@@ -22,12 +22,13 @@ def fill_stripe_form(
     number: 4242_4242_4242_4242, # rubocop:disable Style/NumericLiterals
     expiry_month: 12,
     expiry_year: Time.current.year + 1,
+    date: nil,
```
Comment (waiting-for-dev) [inline spec/support/solidus_stripe/checkout_test_helper.rb:61]: Why do we need it, and how did we only catch it now? :thinking:

File: spec/support/solidus_stripe/checkout_test_helper.rb [hunk 5]
```diff
@@ -77,4 +78,146 @@ def find_stripe_iframe
     expect(fieldset).to have_css('iframe') # trigger waiting if the frame is not yet there
     fieldset.find("iframe")
   end
+
+  def clear_stripe_form
+    %w[number expiry cvc postalCode].each do |name|
+      using_wait_time(10) do
+        within_frame(find_stripe_iframe) do
+          field = find(%{input[name="#{name}"]})
+          field.value.length.times { field.send_keys [:backspace] }
```
Comment (waiting-for-dev) [inline spec/support/solidus_stripe/checkout_test_helper.rb:116]: Isn't it possible to just set `field.value = ''` (probably I'm missing something)?

File: spec/support/solidus_stripe/checkout_test_helper.rb [hunk 6]
```diff
@@ -77,4 +78,146 @@ def find_stripe_iframe
...
+        end
+      end
+    end
+  end
+
+  def submit_payment
+    click_button("Save and Continue")
+  end
+
+  def stripe_payment_method
```
Comment (waiting-for-dev) [inline spec/support/solidus_stripe/checkout_test_helper.rb]: Thoughts about renaming it to `payment_method`. Otherwise, it looks like we're referring to a Stripe payment method resource.

File: spec/support/solidus_stripe/checkout_test_helper.rb [hunk 7]
```diff
@@ -77,4 +78,146 @@ def find_stripe_iframe
...
+    # Memoize the payment method id to avoid fetching it multiple times
+    @stripe_payment_method ||= SolidusStripe::PaymentMethod.first!
+  end
+
+  def create_stripe_payment_method(
```
Comment (waiting-for-dev) [inline spec/support/solidus_stripe/checkout_test_helper.rb]: Same here about confusing `stripe_payment_method`.

File: spec/support/solidus_stripe/checkout_test_helper.rb [hunk 8]
```diff
@@ -77,4 +78,146 @@ def find_stripe_iframe
...
+    setup_intent = SolidusStripe::SetupIntent.find_by(order: order)
+    expect(setup_intent.stripe_intent_id).to be_present
+    intent = setup_intent.stripe_intent
+    expect(intent.status).to eq('succeeded')
+  end
+
+  def expect_payment_intent_to_be_created_successfully
+    payment = Spree::Payment.valid.last
+    expect(payment.payment_method.type).to eq('SolidusStripe::PaymentMethod')
+    intent = payment.payment_method.gateway.request do
```
Comment (waiting-for-dev) [inline spec/support/solidus_stripe/checkout_test_helper.rb]: We can use something similar to setup intents when https://github.com/solidusio/solidus_stripe/pull/217 is merged.

File: spec/support/solidus_stripe/checkout_test_helper.rb [hunk 9]
```diff
@@ -77,4 +78,146 @@ def find_stripe_iframe
...
+
+  def expect_setup_intent_to_be_created_successfully
+    order = Spree::Order.last
+    setup_intent = SolidusStripe::SetupIntent.find_by(order: order)
+    expect(setup_intent.stripe_intent_id).to be_present
+    intent = setup_intent.stripe_intent
+    expect(intent.status).to eq('succeeded')
+  end
+
+  def expect_payment_intent_to_be_created_successfully
```
Comment (waiting-for-dev) [inline spec/support/solidus_stripe/checkout_test_helper.rb]: Thoughts on using a more explicit name, like the `expect_payment_intent_to_be_created_with_required_action` below?

File: spec/support/solidus_stripe/checkout_test_helper.rb [hunk 10]
```diff
@@ -77,4 +78,146 @@ def find_stripe_iframe
...
+  # the "next_action" property of the Stripe PaymentIntent or SetupIntent object.
+
+  # More information on this property can be found in the Stripe API documentation:
+  # - PaymentIntent objects : https://stripe.com/docs/api/payment_intents/object#payment_intent_object-next_action
+  # - SetupIntent objects   : https://stripe.com/docs/api/setup_intents/object#setup_intent_object-next_action
+
+  # Submits a 3D Secure payment authorization
+  def authorize_3d_secure_payment(authenticate: true)
+    find_frame('body > div > iframe') do
+      sleep 1
```
Comment (waiting-for-dev) [inline spec/support/solidus_stripe/checkout_test_helper.rb]: Would `using_waiting_time` work (as used above)?

File: spec/support/solidus_stripe/checkout_test_helper.rb [hunk 11]
```diff
@@ -77,4 +78,146 @@ def find_stripe_iframe
...
+          click_on authenticate ? 'Complete authentication' : 'Fail authentication'
+        end
+      end
+    end
+  end
+
+  # Submits a 3D Secure 2 payment authorization
+  def authorize_3d_secure_2_payment(authenticate: true)
+    find_frame('body > div > iframe') do
+      sleep 1
```
Comment (waiting-for-dev) [inline spec/support/solidus_stripe/checkout_test_helper.rb]: Same about `using_waiting_time`.

File: spec/system/frontend/solidus_stripe/checkout_spec.rb [hunk 12]
```diff
@@ -2,218 +2,302 @@
...
-    expect(order.payments.valid.count).to eq(1)
-    payment.capture!
-    expect_payments_state(order, ['completed'], outstanding: 0)
-    expect(SolidusStripe::PaymentSource.count).to eq(1)
-    expect(SolidusStripe::PaymentSource.last.stripe_payment_method_id).to be_present
-    expect(payment.source).to eq(reusable_source)
-  end
+  # To learn more about setup_future_usage in different contexts:
+  # - SetupIntents   : https://stripe.com/docs/payments/setup-intents#increasing-success-rate-by-specifying-usage
+  # - PaymentIntents : https://stripe.com/docs/payments/payment-intents#future-usage
```
Comment (waiting-for-dev) [inline spec/system/frontend/solidus_stripe/checkout_spec.rb]: Should we move the comment for payment intents above the appropriate block bellow?

File: spec/system/frontend/solidus_stripe/checkout_spec.rb [hunk 13]
```diff
@@ -2,218 +2,302 @@
...
+          expect(page).to have_content('An unexpected error occurred')
+        end
+      end
+    end
   end
 
-  def submit_payment
-    click_button("Save and Continue")
-    expect(page).to have_content("Agree to Terms of Service")
+  private
```
Comment (waiting-for-dev) [inline spec/system/frontend/solidus_stripe/checkout_spec.rb]: I need some clarification. Why the methods below don't go into the checkout helper file?

File: spec/system/frontend/solidus_stripe/checkout_spec.rb [hunk 14]
```diff
@@ -2,218 +2,302 @@
...
-    payment.capture!
-    expect_payments_state(order, ['completed'], outstanding: 0)
-    expect(SolidusStripe::PaymentSource.count).to eq(1)
-    expect(SolidusStripe::PaymentSource.last.stripe_payment_method_id).to be_present
-    expect(payment.source).to eq(reusable_source)
-  end
+  # To learn more about setup_future_usage in different contexts:
+  # - SetupIntents   : https://stripe.com/docs/payments/setup-intents#increasing-success-rate-by-specifying-usage
+  # - PaymentIntents : https://stripe.com/docs/payments/payment-intents#future-usage
+  ['', 'on_session', 'off_session'].each do |setup_future_usage|
```
Comment (waiting-for-dev) [inline spec/system/frontend/solidus_stripe/checkout_spec.rb]: Why `''`? :thinking: It [looks like](https://stripe.com/docs/api/setup_intents/object) `setup_future_usage` might only be `on_session` or `off_session`, but I might be missing something.

File: spec/system/frontend/solidus_stripe/checkout_spec.rb [hunk 15]
```diff
@@ -2,218 +2,302 @@
...
+          fill_stripe_form(number: '4000002760003184')
+
+          # Submit the payment form and authorize the 3D Secure payment
+          submit_payment
+          authorize_3d_secure_payment
+
+          # Expect the setup intent to be created successfully and complete the order
+          expect_setup_intent_to_be_created_successfully
+          complete_order
+          expect(page).to have_content('Your order has been processed successfully')
```
Comment (waiting-for-dev) [inline spec/system/frontend/solidus_stripe/checkout_spec.rb]: This expectation is already covered within the `complete_order` method, probably a leftover :slightly_smiling_face:

File: spec/system/frontend/solidus_stripe/checkout_spec.rb [hunk 16]
```diff
@@ -2,218 +2,302 @@
...
-    expect_payments_state(order, ['completed'], outstanding: 0)
-    expect(SolidusStripe::PaymentSource.count).to eq(1)
-    expect(SolidusStripe::PaymentSource.last.stripe_payment_method_id).to be_present
-    expect(payment.source).to eq(reusable_source)
-  end
+      # incorrect_number
+      fill_stripe_form(number: '4242424242424241')
+      submit_payment
+      using_wait_time(10) do
+        within_frame(find_stripe_iframe) do
```
Comment (waiting-for-dev) [inline spec/system/frontend/solidus_stripe/checkout_spec.rb]: Following the logic used in the rest of the test file, should we extract that to a helper method?

**Review by waiting-for-dev — APPROVED** [review]
Super-clean job, @rainerdema :clap: :clap: I love how it flows.

I left some non-blocking comments & suggestions. Also, to go for honors, there're some inconsistencies in the method names. Most of them perform both flow + expectations, and some of them are named for the expectation (`expect_declined_cards_to_be_notified`), others like a statement(`successfully_creates_a_setup_intent`), and others like an order (`complete_order`).

File: spec/system/frontend/solidus_stripe/checkout_spec.rb (hunk 14 above)
Comment (waiting-for-dev) [reply spec/system/frontend/solidus_stripe/checkout_spec.rb]: I guess it's there to be consumed when using the payment intent flow... could we use it only in that case? Otherwise, we have a redundant test for setup intents.

File: spec/support/solidus_stripe/checkout_test_helper.rb (hunk 4 above)
Comment (waiting-for-dev) [reply spec/support/solidus_stripe/checkout_test_helper.rb:61]: I got it. That's to simulate an error path.

File: spec/support/solidus_stripe/checkout_test_helper.rb (hunk 5 above)
Comment (elia) [reply spec/support/solidus_stripe/checkout_test_helper.rb:116]: I can chime in on this one, as it's coming from a previous helper. Basically this is the result of trial and error, and was the most stable way of filling in fields in the stripe iframe, I believe partly because of the JavaScript they run whenever a key is typed, but it's just a hypothesis.

Comment (rainerdema) [reply spec/support/solidus_stripe/checkout_test_helper.rb:116]: I can confirm that I have tried several solutions, but this is the most stable one I have found.

File: spec/support/solidus_stripe/checkout_test_helper.rb (hunk 10 above)
Comment (rainerdema) [reply spec/support/solidus_stripe/checkout_test_helper.rb]: I'll make a few tries, but probably yes (and it's already in the `find_frame` helper method)

File: spec/system/frontend/solidus_stripe/checkout_spec.rb (hunk 13 above)
Comment (rainerdema) [reply spec/system/frontend/solidus_stripe/checkout_spec.rb]: For no particular reason 😄 Upon re-reading the specs, I noticed that these methods were the primary ones most used for constructing the various test blocks. As a result, I decided to keep them easily accessible within the `checkout_spec.rb` file. Anyway, I'll move them all to the dedicated helper for consistency 👍

File: spec/system/frontend/solidus_stripe/checkout_spec.rb (hunk 3 above)
Comment (rainerdema) [reply spec/system/frontend/solidus_stripe/checkout_spec.rb]: Great idea! 👏  I've made the changes:
https://github.com/solidusio/solidus_stripe/blob/759cb54b80ce518a1f97828d29aa4677630d07de/spec/system/frontend/solidus_stripe/checkout_spec.rb#L66

This is the result:
Screenshot 2023-03-15 at 19 19 28

Let me know if you were considering adding the `pending` status in the header of the "it" block instead (With the message, it appeared to be more comprehensive to me 🤔).

File: spec/system/frontend/solidus_stripe/checkout_spec.rb (hunk 16 above)
Comment (rainerdema) [reply spec/system/frontend/solidus_stripe/checkout_spec.rb]: Great suggestion! I've moved this part to a dedicated method:
https://github.com/solidusio/solidus_stripe/blob/759cb54b80ce518a1f97828d29aa4677630d07de/spec/support/solidus_stripe/checkout_test_helper.rb#L242-L259

**Comment by rainerdema** [comment]
@elia @kennyadsl @waiting-for-dev 👋 Thanks everyone for the reviews!
I have looked over your suggestions and made the necessary revisions.

Here are a few considerations:
- Considering the number of methods in the [checkout_test_helper.rb](https://github.com/solidusio/solidus_stripe/blob/90c4482934057b90219afeed84290ff13305fdf9/spec/support/solidus_stripe/checkout_test_helper.rb), I have reorganized and grouped them as efficiently as possible.
- I'm waiting for the merge of #217 to integrate the retrieve methods into the helper before merging (https://github.com/solidusio/solidus_stripe/pull/215#discussion_r1136709134)
- I have removed the static `off_session` setting from the setup intent creation to ensure that the tests correctly cover all configuration flows: https://github.com/solidusio/solidus_stripe/pull/215/commits/90c4482934057b90219afeed84290ff13305fdf9
Let me know if we're ok with that 👍

**Comment by rainerdema** [comment]
> there're some inconsistencies in the method names. Most of them perform both flow + expectations, and some of them are named for the expectation (expect_declined_cards_to_be_notified), others like a statement(successfully_creates_a_setup_intent), and others like an order (complete_order).

@waiting-for-dev Yes, I agree with you that In `expect_declined_cards_to_be_notified`, `successfully_creates_a_setup_intent` and `complete_order` for example, there are expectation instructions that appear to be mixed in with flow instructions.
I'm open to suggestions in case the method nomenclature can be better improved 👍

However, most of the time, these checks are necessary to ensure that the subsequent commands are executed at the appropriate time.
Let me know if this makes sense and if it addresses your concerns.

**Comment by waiting-for-dev** [comment]
> However, most of the time, these checks are necessary to ensure that the subsequent commands are executed at the appropriate time.
> Let me know if this makes sense and if it addresses your concerns.

Yeah, I'm not suggesting removing the checks. Just using the same naming convention. For instance, for the three examples:

- `declined_cards_are_notified`
- `successfully_creates_a_setup_intent`
- `completes_order`

I'm picking the statement format because it's the more neutral between `expect_`, which looks like it only contains RSpec expectation`, and the imperative, which looks like it only contains flow instructions.

But leaving it to your chose!

> I'm waiting for the merge of https://github.com/solidusio/solidus_stripe/pull/217 to integrate the retrieve methods into the helper before merging (https://github.com/solidusio/solidus_stripe/pull/215#discussion_r1136709134)

Just merged! :slightly_smiling_face:
